### DISCUSSION TITLE
RavenDB support of ReadUncommitted

### DISCUSSION THREAD

**Discussion by ashokgudur** [ravendb/ravendb#13914] [General]
Hi,

Is RavenDB supports ReadUncommitted?
I have huge number of documents ( approx. 3000 invoices) which comes from an external source. I would need to conditionally either insert or update the doc.  The conditional logic involves reading by a query (in the same session) to see if it already exists, and then make some changes and update. If not exist insert new one, but again it can be a updated in next iteration in the same same session.

Finally I should make sure the whole operation either saved or rejected (transactional).

Am I correct here?

Thanks.

**Comment by ayende** [comment]
RavenDB doesn't have the notion of read uncommitted. However, what you are describing is not _that_. You are talking about reading your own modifications inside the same transaction, which is absolutely supported.

The easiest way to do that is using `Patch` operations. In particular, see `session.Advanced.AddOrPatch();`

That would either save the entity, if it is missing, or apply the patch to it. Here is an example of its usage:

```
session.Advanced.AddOrPatch<User, DateTime>(
    id,
    new User {
        FirstName = "Hibernating", 
        LastName = "Rhinos", 
        LoginTimes = new List<DateTime>{DateTime.UtcNow}},
    x => x.LoginTimes, 
    u => u.Add(new DateTime(1993,09,12),new DateTime(2000,01,01)));
```

You can also do more complex stuff if you directly use the API and apply logic directly:

```
session.Advanced.Defer(new PatchCommandData("items/first", null,
    new Raven.Client.Documents.Operations.PatchRequest
    {
        Script = @"
var self = this;
Object.keys(this.Trie).filter(function(key) {
if(self.Trie[key].PageId === args.PageId) {
delete self.Trie[key];
}
});
",
        Values =
        {
            ["PageId"] = "home",
        },
    })
{
    CreateIfMissing = newInstance
});
```

**Comment by ashokgudur** [comment]
OK.
thanks
